packages/library/src/rpc-websocket-connection-sharding.ts

```diff
@@ -0,0 +1,122 @@
+}
+
+export function getWebSocketTransportWithConnectionSharding({ getShard, transport }: Config): IRpcWebSocketTransport {
+    const cache = new Map<CacheKey, CacheEntry>();
+    function updateCache(shardKey: CacheKey, updater: (currentCacheEntry: CacheEntry) => CacheEntry) {
+        const currentCacheEntry = cache.get(shardKey);
+        if (!currentCacheEntry) {
+            // TODO: Coded error.
+            throw new Error(`Found no cache entry for connection with shard key \`${shardKey.toString()}\``);
+        }
+        const nextCacheEntry = updater(currentCacheEntry);
+        cache.set(shardKey, nextCacheEntry);
+        return nextCacheEntry;
+    }
+    return async (...args) => {
+        const { payload, signal, ...rest } = args[0];
+        const shardKey = getShard ? getShard(payload) : NULL_SHARD_CACHE_KEY;
+        function cleanup() {
+            cache.delete(shardKey);
+            signal.removeEventListener('abort', handleAbort);
+        }
+        function handleAbort() {
+            if (cache.get(shardKey)?.purgeScheduled !== true) {
+                updateCache(shardKey, currentCacheEntry => ({
+                    ...currentCacheEntry,
+                    purgeScheduled: true,
+                }));
+                globalThis.queueMicrotask(() => {
+                    const cacheEntryAtEndOfRunloop = cache.get(shardKey);
+                    if (!cacheEntryAtEndOfRunloop) {
+                        return;
+                    }
+                    if (cacheEntryAtEndOfRunloop.referenceCount === 0) {
+                        cacheEntry.abortController.abort();
+                        cleanup();
+                    }
+                });
+            }
+            const cacheEntry = updateCache(shardKey, currentCacheEntry => ({
+                ...currentCacheEntry,
+                referenceCount: currentCacheEntry.referenceCount - 1,
+            }));
+        }
+        signal.addEventListener('abort', handleAbort);
+        try {
+            const cacheEntry = cache.get(shardKey);
+            if (!cacheEntry) {
+                const connectionAbortController = new AbortController();
+                const newConnectionPromise = transport({
+                    payload,
+                    signal: connectionAbortController.signal,
+                    ...rest,
+                });
+                const newCacheEntry = {
+                    abortController: connectionAbortController,
+                    connection: newConnectionPromise,
+                    purgeScheduled: false,
+                    referenceCount: 1,
+                };
+                cache.set(shardKey, newCacheEntry);
+                const newConnection = await newConnectionPromise;
+                registerIterableCleanup(newConnection, cleanup);
+                updateCache(shardKey, currentCacheEntry => ({
+                    ...currentCacheEntry,
+                    connection: newConnection,
+                }));
```

Comment:
How come we store `newConnection` instead of `newConnectionPromise` here like we do in `newCacheEntry`, and simplify that type to just storing the `Awaited<...>`. Or similarly why don't we await in in the other case and simplify the type to `ReturnType<IRpcWebSocketTransport>`?